src/main/java/com/rultor/agents/github/PreviousReleases.java

```diff
@@ -0,0 +1,140 @@
+     */
+    private final transient Version version;
+
+    /**
+     * Previously released tags.
+     */
+    private final transient Releases priors;
+
+    /**
+     * Default constructor.
+     * @param ver The proposed tag, wrapped in a Version object
+     * @param previous A list of previously released tags.
+     */
+    public PreviousReleases(final Version ver, final Releases previous) {
+        this.version = ver;
+        this.priors = previous;
+    }
+
+    /**
+     * Is the proposed tag true, with respect to previous released tags. For
+     * example, if there are previous releases of [0.2,0.5,0.7], and the
+     * proposed tag (version) is .6, this will return false. If the proposed
+     * tag is .8 it will return true.
+     * @return True if the release is valid
+     */
+    public boolean isValid() {
+        final DefaultArtifactVersion latest = latest();
+        return new DefaultArtifactVersion(this.version.toString())
+            .compareTo(latest) == 1;
+    }
+
+    /**
+     * Returns the latest release. If there are no releases, returns "0".
+     * @return The latest release, otherwise "0" if there are no releases.
+     */
+    public DefaultArtifactVersion latest() {
+        final DefaultArtifactVersion latest;
+        final List<DefaultArtifactVersion> previous = this.transformVersions();
+        if (previous.isEmpty()) {
+            latest = new DefaultArtifactVersion("0");
+        } else {
+            latest = Collections.max(previous);
+        }
+        return latest;
+    }
+
+    /**
+     * Transforms versionsFrom Release to DefaultArtificatVersion and filters
+     * invalid version numbers. For example in these versions,
+     * ["1.0", "2.0", "3.0-b"], "3.0-b" is just ignore, therefore version "2.0"
+     * is the max.
+     * @return All prior releases wrapped in a DefaultArtifactVersion
+     */
+    private List<DefaultArtifactVersion> transformVersions() {
+        return FluentIterable
+            .from(this.priors.iterate())
+            .transform(new ReleaseToVersion())
+            .filter(new ValidVersion())
+            .toList();
+    }
+
+    private final class ValidVersion
+        implements Predicate<DefaultArtifactVersion> {
+        @Override
+        public boolean apply(final DefaultArtifactVersion ver) {
+            return new Version(ver.toString()).isValid();
+        }
+    }
+
+    private final class ReleaseToVersion
+        implements Function<Release, DefaultArtifactVersion> {
+        @Override
+        @SuppressWarnings("PMD.OnlyOneReturn")
+        public DefaultArtifactVersion apply(final Release release) {
+            final Release.Smart rel = new Release.Smart(release);
+            try {
+                final String tag = rel.tag();
+                return new DefaultArtifactVersion(tag);
+            } catch (final IOException exception) {
+                Logger.error(this, "IOException caused from rel.tag()");
+                return null;
```

Comment:
@yegor256 

I think failing fast is the best practice. I originally had it like this but PMD complained, so I changed it to return null. Funny enough, it still complained so I ignored that suppression without thinking to ignore the previous and make it fail fast. This was the original code:

```java
    private final class ReleaseToVersion
        implements Function<Release, DefaultArtifactVersion> {
        @Override
        public DefaultArtifactVersion apply(final Release release) {
            final Release.Smart rel = new Release.Smart(release);
            try {
                final String tag = rel.tag();
                return new DefaultArtifactVersion(tag);
            } catch (final IOException exception) {
                Logger.error(this, "IOException caused from rel.tag()");
                throw new RuntimeException(exception);
            }
        }
    }
```

Should I change it back to this version, with `throw new RuntimeException(e);` ? It's not possible that the client code can recover from the exception.